### PR TITLE
add /clusters/{cluster_id}/cves endpoint

### PR DESCRIPTION
**Pull request by yungbender**
- Refactored gin context helper functions to `context.go`
- Added cluster-cves endpoint and added  **complete** openapi docs

## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [x] Input Validation
- [x] Output Encoding
- [x] Authentication and Password Management
- [x] Session Management
- [x] Access Control
- [x] Cryptographic Practices
- [x] Error Handling and Logging
- [x] Data Protection
- [x] Communication Security
- [x] System Configuration
- [x] Database Security
- [x] File Management
- [x] Memory Management
- [x] General Coding Practices

### CODE REVIEW

File: manager/controllers/clusters/cves.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+// @Param limit 	 	  query int	   	 false "limit per page"		  example(10)
+// @Param offset	 	  query int	   	 false "page offset"		  example(10)
+// @Param published   	  query []string false "CVE publish date: (from date),(to date)" collectionFormat(multi) collectionFormat(csv) minItems(2) maxItems(2)
+// @Param severity   	  query []string false "severity name" enums(NotSet,None,Low,Medium,Moderate,Important,High,Critical)
+// @Param cvss_score 	  query []number false "CVSS score of CVE: (from float),(to float)" collectionFormat(multi) collectionFormat(csv) minItems(2) maxItems(2)
+// @Param affected_images query []bool 	 false "minimal count of affected images: (1 or more),(none)" collectionFormat(multi) collectionFormat(csv) minItems(2) maxItems(2)
+// @router /clusters/{cluster_id}/cves [get]
+// @success 200 {object} base.Response{data=GetClusterCvesSelect}
+// @failure 404 {object} base.Error "cluster does not exist"
+// @failure 500 {object} base.Error
```
Comment (blayson) [inline manager/controllers/clusters/cves.go:69]: There is missed 400 error code (bad request)

File: manager/controllers/clusters/cves.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+//
+// @id GetClusterCves
+// @summary List of CVEs affecting/nonaffecting single cluster
+// @security RhIdentity || BasicAuth
+// @Tags clusters
+// @description Endpoint returning CVEs affecting/nonaffecting the given single cluster
+// @accept */*
+// @produce json
+// @Param cluster_id 	  path  string 	 true  "Cluster ID"
+// @Param sort		 	  query []string false "column for sort"	  collectionFormat(multi) collectionFormat(csv)
```
Comment (blayson) [inline manager/controllers/clusters/cves.go:58]: Please fix indentation here

**Review by blayson — CHANGES_REQUESTED** [review]

File: manager/controllers/clusters/cves.go (hunk 2 above)
Comment (yungbender) [reply manager/controllers/clusters/cves.go]: fixed

File: manager/base/context.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+			return f
+		}
+	}
+	return map[string]Filter{}
+}
+
+// GetUUID gets raw value from gin context and returns an parsed UUID (or error)
+func GetParamUUID(ctx *gin.Context, key string) (uuid.UUID, error) {
+	rawVal := ctx.Param(key)
+	fmt.Print(rawVal)
```
Comment (jdobes) [inline manager/base/context.go:24]: remove

File: manager/controllers/clusters/cves.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+	subquery := c.Conn.Table("cluster_image").
+		Select("cluster_image.image_id").
+		Joins("JOIN cluster ON cluster_image.cluster_id = cluster.id").
+		Where("cluster.account_id = ? AND cluster.uuid = ?", accountID, clusterID.String())
+
+	return c.Conn.Table("cve").
+		Select(`cve.cvss2_score, cve.cvss3_score, cve.description, cve.severity,
+			cve.public_date, cve.name,
+			COUNT(DISTINCT cluster_image.image_id) as images_exposed`).
+		Joins("LEFT JOIN image_cve ON cve.id = image_cve.cve_id").
```
Comment (jdobes) [inline manager/controllers/clusters/cves.go:138]: shouldn't be INNER join to return only affected CVEs?

**Review by jdobes — CHANGES_REQUESTED** [review]

File: manager/controllers/clusters/cves.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+// @security RhIdentity || BasicAuth
+// @Tags clusters
+// @description Endpoint returning CVEs affecting/nonaffecting the given single cluster
+// @accept */*
+// @produce json
+// @Param cluster_id      path  string   true  "Cluster ID"
+// @Param sort            query []string false "column for sort"	  collectionFormat(multi) collectionFormat(csv)
+// @Param search          query string   false "cve name/desc search" example(CVE-2021-)
+// @Param limit           query int	   	 false "limit per page"		  example(10)
+// @Param offset          query int	   	 false "page offset"		  example(10)
```
Comment (blayson) [inline manager/controllers/clusters/cves.go:61]: there is still two lines with same issue, probably it's because of tabs :D

**Review by blayson — CHANGES_REQUESTED** [review]

File: manager/base/context.go (hunk 3 above)
Comment (yungbender) [reply manager/base/context.go]: fixed

File: manager/controllers/clusters/cves.go (hunk 4 above)
Comment (yungbender) [reply manager/controllers/clusters/cves.go:138]: We want to support `affected_images` filter, thus left join if customer wants to see cves with 0 affected images.

File: manager/controllers/clusters/cves.go (hunk 5 above)
Comment (yungbender) [reply manager/controllers/clusters/cves.go]: should be better now

File: manager/controllers/clusters/cves.go (hunk 4 above)
Comment (yungbender) [reply manager/controllers/clusters/cves.go]: Reworked it so CVEs with 0 images exposed are not returned and all and removed the filter from openapi.

**Review by jdobes — APPROVED** [review]

**Pull request merged**